### PR TITLE
Fixed #34012 -- Made QuerySet.order_by() apply transforms on related fields for models with Meta.ordering.

### PR DESCRIPTION
**Pull request by shangxiao**
Hi Carlton/Mariusz,

I had a look through the ordering code and it seems safe to assume that anything of the form `order_by(related__something)` must assume "something" is a field and not a transform. If we look at `find_ordering_name()` [it ignores any transforms after a relation field and starts looking at the `Meta.ordering` of that relation](https://github.com/django/django/blob/3ba7f2e9069c54db6d6d9d2fd1945b2dbc935d9c/django/db/models/sql/compiler.py#L1002-L1010) after calling `_setup_joins()`.

The only thing I'm unsure of is whether this could impact any code not dealing with ordering.

Note: Btw is someone able to confirm transforms on `order_by()` is still a thing? I just tried it and it doesn't appear to work using a simple `order_by('char_field__length')` 🤔

### CODE REVIEW

**Comment by felixxm** [comment]
Thanks for this patch, however I don't think it's the right approach, especially after 10178197d57476f69688d4535e550a1ea3a5eac5. In Django 4.2+ (the current `main` branch) you can register lookups/transforms on field instances, e.g.:
```diff
diff --git a/tests/queries/tests.py b/tests/queries/tests.py
index 1bd72dd8b8..65164f73c1 100644
--- a/tests/queries/tests.py
+++ b/tests/queries/tests.py
@@ -1621,6 +1621,19 @@ class Queries4Tests(TestCase):
             date_obj,
         )
 
+    def test_related_transform(self):
+        from django.db.models.functions import ExtractYear
+        from django.test.utils import register_lookup
+
+        date_obj = DateTimePK.objects.create()
+        extra_obj = ExtraInfo.objects.create(info="extra", date=date_obj)
+        fk_field = ExtraInfo._meta.get_field("date")
+        with register_lookup(fk_field, ExtractYear):
+            self.assertIn(
+                "django_datetime_extract",
+                str(ExtraInfo.objects.order_by("date__year").query),
+            )
+
     def test_ticket10181(self):
         # Avoid raising an EmptyResultSet if an inner query is probably
         # empty (and hence, not executed).
```
`test_related_transform` works on SQLite without this patch. However with #16070 it crashes:
```
django.core.exceptions.FieldError: Cannot resolve keyword 'year' into field. Choices are: date, extrainfo
```

**Comment by shangxiao** [comment]
@felixxm ah so if lookups on related fields is a thing then yeah this is the wrong approach 👍 If there's an alternative that works the test above that uses a datetime fk is worth adding to the test suite.

**Comment by felixxm** [comment]
See #16074.

**Comment by shangxiao** [comment]
@felixxm I've pushed another demo commit with an alternative way to check for transforms - by just returning the list of transforms from setup_joins().  I think I like this approach better but it does change the function return type which impacts more code.

A check is failing on a "forever test" - not sure what that's about.

**Comment by shangxiao** [comment]
@felixxm I've pushed a 3rd demo commit with another alternative which may be the best choice so far: Simply inspecting the transform function, which is partially applied to do the lookups, for a lookup name is good enough to detect whether there are lookups: d6204bd1136417131e360685eab1640d1f990a28

I'd personally choose this as there's minimal impact on the codebase.

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -1803,6 +1803,7 @@ def final_transformer(field, alias):
                 # resolved into fields.
                 transforms = names[pivot:]
                 break
+
```
Comment (felixxm) [inline django/db/models/sql/query.py:1806]: Revert unrelated change.

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -999,14 +999,20 @@ def find_ordering_name(
...
         # append the default ordering for that model unless it is the pk
-        # shortcut or the attribute name of the field that is specified.
+        # shortcut or the attribute name of the field that is specified or
+        # there are lookups to process.
         if (
             field.is_relation
             and opts.ordering
             and getattr(field, "attname", None) != pieces[-1]
             and name != "pk"
+            and not has_lookups
```
Comment (felixxm) [inline django/db/models/sql/compiler.py:1015]: What do you think about assigning `has_lookups` to the `transform_function` and be more explicit here?
```suggestion
            and not getattr(transform_function, "has_lookups", False)
```
and
```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index 7e43ec254d..58ccb8f923 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -1820,6 +1820,7 @@ class Query(BaseExpression):
             final_transformer = functools.partial(
                 transform, name=name, previous=final_transformer
             )
+            final_transformer.has_lookups = True
         # Then, add the path to the query's joins. Note that we can't trim
         # joins at this stage - we will need the information about join type
         # of the trimmed joins.
```

**Review by felixxm — COMMENTED** [review]
@shangxiao Thanks for updates :+1:

File: django/db/models/sql/compiler.py (hunk 2 above)
Comment (shangxiao) [reply django/db/models/sql/compiler.py:1015]: @felixxm Yep that's a good way to do it too - 1 function is created per invocation so it would work 👍. Explicit, readable & minimal api impact. The one con is that it could be surprising to future maintainers if function attributes aren't commonly done like this in this community? 🤔

If you're happy to do this I can squash everything down - or if you wanted to do it 😊🤷‍♂️

Comment (felixxm) [reply django/db/models/sql/compiler.py:1015]: As far as I'm aware, assigning variables to functions is not sth unusual :thinking: 

> If you're happy to do this I can squash everything down - or if you wanted to do it

Please do. Also, `has_transforms` is probably more accurate than `has_lookups`.
